### PR TITLE
Add Cilium

### PR DESCRIPTION
**Pull request by errordeveloper**
This PR adds steps and workflows to enable testing with [Cilium](https://github.com/cilium/cilium).

Steps and workflows are defined aws, azure & gcp. As Cilium requires additional ports, only an azure periodic is being added here for 4.8. Peridics for gcp & aws should be added in the future, but something will have to be done about Cilium ports.

### CODE REVIEW

**Comment by vrutkovs** [comment]
_copying from slack thread_

Templates are deprecates, so ideally this would be converted into steps.
see https://docs.ci.openshift.org/docs/architecture/step-registry/

I think the best here would be:
* Create a step similar to https://github.com/openshift/release/tree/master/ci-operator/step-registry/ipi/install/hosted-loki which unpacks Cilium manifests to ${SHARED_DIR} and sets network: none in install-config.yaml
* an ipi-aws-cilium chain which uses that step
* a new openshift-e2e-cilium-aws workflow

In order to test this you could add a periodic job which runs this workflow

**Comment by errordeveloper** [comment]
@vrutkovs I had a look around and figured that'd base this on structure similar to e.g. OVN rather then Loki, PTAL!

**Review by vrutkovs — CHANGES_REQUESTED** [review]
This looks good!

Please add this to the periodics (similar to https://github.com/openshift/release/pull/16278) so that rehearsals would give it a try

File: ci-operator/step-registry/cilium/OWNERS
Comment (vrutkovs) [inline ci-operator/step-registry/cilium/OWNERS:2]:
```suggestion
- errordeveloper
- smarterclayton
- wking
- stevekuznetsov
- vrutkovs
- abhinavdahiya
- deads2k
```
for more eyes on reviews (same for other OWNERS files there)

**Comment by vrutkovs** [comment]
AWS install is pretty strict about open ports, lets try `azure` instead?

**Comment by errordeveloper** [comment]
@vrutkovs it definitely makes sense to prioritise Azure and make sure things work overall! I'll happily leave UPI and opening ports in GCP and AWS for another PR ;)

File: ci-operator/step-registry/cilium/conf/cilium-conf-ref.metadata.json
Comment (vrutkovs) [inline ci-operator/step-registry/cilium/conf/cilium-conf-ref.metadata.json:2]: Mix of tabs and spaces.

Use `make update` to regenerate these files

File: ci-operator/step-registry/openshift/e2e/gcp/cilium/default/openshift-e2e-gcp-cilium-default-workflow.yaml
Comment (vrutkovs) [inline ci-operator/step-registry/openshift/e2e/gcp/cilium/default/openshift-e2e-gcp-cilium-default-workflow.yaml:2]:
```suggestion
  as: openshift-e2e-gcp-cilium
```
I'd rather remove `default` and add `openshift-e2e-gcp-cilium-someotherconfig` later

**Comment by vrutkovs** [comment]
Lets add a new periodic (see https://github.com/openshift/release/pull/17519/files#diff-53da59c71bd31de3834afc38a2ccc62bf85595cb1546630ce3a10d4bfe6fe695) to have this rehearsed in CI

File: ci-operator/step-registry/cilium/conf/cilium-conf-ref.metadata.json
Comment (errordeveloper) [reply ci-operator/step-registry/cilium/conf/cilium-conf-ref.metadata.json:2]: Ah, didn't realise these were generated... thanks!

**Comment by errordeveloper** [comment]
@vrutkovs once I add the periodic, will only run once the changes are merged, or it can be rehearsed from the PR branch somehow?

**Comment by vrutkovs** [comment]
> @vrutkovs once I add the periodic, will only run once the changes are merged, or it can be rehearsed from the PR branch somehow?

CI bot would add `rehearse-...` jobs to this PR. We won't merge the PR until it passes

File: ci-operator/step-registry/cilium/conf/cilium-conf-commands.sh
Comment (vrutkovs) [inline ci-operator/step-registry/cilium/conf/cilium-conf-commands.sh:9]: If I understood correctly this effectively disables cluster network operator by setting some unknown network type (so that we could setup network from manifests later)?

Lets make this step more generic - `ci-operator/step-registry/network/conf/disable-cno/...` and set `networkType: none` (this would allow us to reuse this step for other CNIs)

Comment (errordeveloper) [reply ci-operator/step-registry/cilium/conf/cilium-conf-commands.sh:9]: Good idea, I will do this.

File: ci-operator/step-registry/openshift/e2e/gcp/cilium/default/openshift-e2e-gcp-cilium-default-workflow.yaml
Comment (errordeveloper) [reply ci-operator/step-registry/openshift/e2e/gcp/cilium/default/openshift-e2e-gcp-cilium-default-workflow.yaml:2]: I was just doing this for consistency, but now your version makes more sense, as there there is `network-conf-disable-cno` instead of `cilium-conf`.

File: ci-operator/config/openshift/release/openshift-release-master__ci-4.8.yaml
Comment (vrutkovs) [inline ci-operator/config/openshift/release/openshift-release-master__ci-4.8.yaml:107]: I'd recommend starting with default `SIZE_VARIANT`. `compact` means 3 masters and 0 workers, so chances are some components might not be able to roll out, causing noise in Cilium jobs

Comment (errordeveloper) [reply ci-operator/config/openshift/release/openshift-release-master__ci-4.8.yaml:107]: Ah, I see

File: ci-operator/step-registry/cilium/conf/cilium-conf-commands.sh
Comment (vrutkovs) [inline ci-operator/step-registry/cilium/conf/cilium-conf-commands.sh:10]: Not sure if its necessary - `cluster-network-02-config.yml` already has similar contents for config.openshift.io/v1 Network.

File: ci-operator/step-registry/cilium/conf/cilium-conf-commands.sh
Comment (vrutkovs) [inline ci-operator/step-registry/cilium/conf/cilium-conf-commands.sh:29]: I don't think that worked, at least I don't see any new manifests in `rendered-assets/openshift/manifests`.

Lets add some more debug options (it won't affect the result) - i.e. `ls -la ${SHARED_DIR}` and some more verbosity on the source of the manifests - version, URL, curl output etc.?

File: ci-operator/step-registry/cilium/conf/cilium-conf-commands.sh
Comment (errordeveloper) [reply ci-operator/step-registry/cilium/conf/cilium-conf-commands.sh:10]: I was staying on the explicit side of things... OVN workflow seem to do this, so I figured I'll do a similar thing here.

File: ci-operator/step-registry/cilium/conf/cilium-conf-commands.sh
Comment (errordeveloper) [reply ci-operator/step-registry/cilium/conf/cilium-conf-commands.sh:29]: I have definitely tested this locally quite a few times, will add `ls -la` for now... I am not sure if `${SHARED_DIR}` points at manifest, but we will find out now.

**Comment by errordeveloper** [comment]
Ah, looks like `$SHARED_DIR` is the where `install-config.yaml` is and `ipi-install-install-commands.sh` looks of `manifest_*.yml` and `manifest_*.yaml` in there. That is a bit odd, I'd expect more user-centric flow with  `openshift-install create manifests` in a dedicated step... Anyway, I know how to fix this now, will look tomorrow.

**Comment by errordeveloper** [comment]
Looks like install was successful already, let's see if the tests pass... I wonder if it might make sense to focus on networking test, wdyt @vrutkovs?

**Comment by vrutkovs** [comment]
>I wonder if it might make sense to focus on networking test

I don't think we have a special networking-focused test suite (@trozet might know), so minimal conformance is fine for now (we'll expand this to serial later)

**Comment by errordeveloper** [comment]
@vrutkovs I was thinking of `FOCUS=\[sig-network\]`.

**Comment by vrutkovs** [comment]
>The HAProxy router should be able to connect to a service that is idled because a GET on the route will unidle it

seems suspicious, I suppose `networkType: none` is not expected to pass idle tests?

>Managed cluster should only include cluster daemonsets that have maxUnavailable update of 10 or 33 percent

I think its fair to skip this one (ideally upstream manifests would use 33% though)


Lets have another round
/retest

**Comment by errordeveloper** [comment]
> > The HAProxy router should be able to connect to a service that is idled because a GET on the route will unidle it
> 
> seems suspicious, I suppose `networkType: none` is not expected to pass idle tests?

Not sure about this one, I'll have to have a closer look. What's the best way to reproduce this in my own cluster?
 
> > Managed cluster should only include cluster daemonsets that have maxUnavailable update of 10 or 33 percent
> 
> I think its fair to skip this one (ideally upstream manifests would use 33% though)

This one is actually tripping on openshift-kube-proxy, perhaps it's passing normally when kube-proxy is disabled with OVN?

**Comment by errordeveloper** [comment]
I'm not sure why on the second run quite a few`[sig-storage]` tests are failing. As for `[sig-network]`, looks like same failures have re-occurred in both runs, so I'm very much inclined to accept these need investigating with Cilium team.

**Comment by vrutkovs** [comment]
Seems storage tests are failing due to issues on the nodes, likely due to `Bug 1812261: iptables is segfaulting`

/retest

**Comment by vrutkovs** [comment]
Failure analysis:
>[sig-network-edge][Conformance][Area:Networking][Feature:Router] The HAProxy router should be able to connect to a service that is idled because a GET on the route will unidle it

Should be skipped, iiuc custom CNI's may not support idling/unidling

> [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client should deny ingress access to updated pod

AFAIK Cilium uses their own NetworkPolicy objects? Should be skipped probably

> Pods cannot access the /config/master API endpoint

Most likely its covered by NetworkPolicy, so see above

Other failures look like flakes to me

**Comment by errordeveloper** [comment]
@vrutkovs  thanks for looking into this!

> Failure analysis:
> 
> > [sig-network-edge][Conformance][Area:Networking][Feature:Router] The HAProxy router should be able to connect to a service that is idled because a GET on the route will unidle it
> 
> Should be skipped, iiuc custom CNI's may not support idling/unidling

I think this is indeed possible, but is this an OpenShift feature on an upstream one? The `[sig-network-edge]` tag seems to suggest that it's something coming from upstream, but I don't really know... 

> > [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client should deny ingress access to updated pod
> 
> AFAIK Cilium uses their own NetworkPolicy objects? Should be skipped probably

Cilium supports `NetworkPolicy` objects as well as own `CiliumNetworkPolicy` CRs, and which passes the upstream tests except for a few that already been skipped in this PR. Is this an upstream test or an OpenShift-specific one?

> > Pods cannot access the /config/master API endpoint
> 
> Most likely its covered by NetworkPolicy, so see above
> 
> Other failures look like flakes to me

So this will be something to look into, unless it's something OpenShift-specific that Cilium doesn't implement yet...

**Comment by errordeveloper** [comment]
@vrutkovs never mind, I found source references in the output and can see which tests are upstream and which ones are OpenShift ones.

**Comment by errordeveloper** [comment]
Just to summaries my finding:
```
[sig-network-edge][Conformance][Area:Networking][Feature:Router] The HAProxy router should be able to connect to a service that is idled because a GET on the route will unidle it [Skipped:Disconnected] [Suite:openshift/conformance/parallel/minimal]
```
This an OpenShift-specific test, we will need to determine what need to be done to make this pass with Cilium, it's possible there is some annotation that Cilium needs to be able to recognise or something else of similar nature.

```
[sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client should deny ingress access to updated pod [Feature:NetworkPolicy] [Skipped:Network/OpenShiftSDN/Multitenant] [Suite:openshift/conformance/parallel] [Suite:k8s]
```
This one is an upstream tests, we will need to investigate why it doesn't pass.
```
[sig-cluster-lifecycle] Pods cannot access the /config/master API endpoint [Skipped:Disconnected] [Suite:openshift/conformance/parallel] 
```
This is an OpenShift-specific test, it's not clear yet why is it failing and that needs more investigation.
```
[sig-network] Services should be rejected when no endpoints exist [Skipped:ibmcloud] [Suite:openshift/conformance/parallel] [Suite:k8s]
```
This one is another upstream test that needs investigating.

File: ci-operator/step-registry/openshift/e2e/azure/cilium/openshift-e2e-azure-cilium-workflow.yaml
Comment (errordeveloper) [inline ci-operator/step-registry/openshift/e2e/azure/cilium/openshift-e2e-azure-cilium-workflow.yaml:20]: @vrutkovs this doesn't seem to work, I thought it did in first iteration and kept using same syntax...

Comment (errordeveloper) [reply ci-operator/step-registry/openshift/e2e/azure/cilium/openshift-e2e-azure-cilium-workflow.yaml:20]: hm... maybe I was looking at wrong results.

**Comment by errordeveloper** [comment]
Another one just came up:
```
[sig-network] network isolation when using a plugin that does not isolate namespaces by default should allow communication between pods in different namespaces on different nodes
 ```

**Comment by errordeveloper** [comment]
Closed accidentally, re-opening.

**Comment by errordeveloper** [comment]
> Few more just came up:
> 
> ```
> [sig-network] network isolation when using a plugin that does not isolate namespaces by default should allow communication between pods in different namespaces on different nodes
> ```
> 
> This is an OpenShift test.
> 
> ```
> [sig-network] Proxy version v1 should proxy logs on node with explicit kubelet port using proxy subresource [Suite:openshift/conformance/parallel] [Suite:k8s] expand_more
> ```
> 
> This is an upstream test.
> 
> ```
>  [sig-network] Networking Granular Checks: Services should function for endpoint-Service: udp [Suite:openshift/conformance/parallel] [Suite:k8s]
> ```
> 
> This one looks like it's upstream also.

Looks like these are flakes, neither have failed [on the last attempt](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/17380/rehearse-17380-periodic-ci-openshift-release-master-ci-4.8-e2e-azure-cilium/1381648548273065984).

**Comment by errordeveloper** [comment]
Looks like after renaming there is old rehearsal job still lurking around: `ci/rehearse/periodic-ci-openshift-release-master-ci-4.8-e2e-azure-cilium-compact`...

**Comment by errordeveloper** [comment]
@vrutkovs I think this is actually good now, how do we go about the flakes? I'll squash the commits also.

**Comment by vrutkovs** [comment]
I think we're good with leaving the flakes for now - we'll squash those after a few rounds of tests.

Lets update the description - why Azure, what this PR does (adds steps / workflows and a single periodic for 4.8) along with future plans.

/cc @deads2k @smarterclayton for review

**Comment by errordeveloper** [comment]
Looks like only one failure in the latest run:
```
[sig-arch] Managed cluster should have no crashlooping pods in core namespaces over four minutes [Suite:openshift/conformance/parallel]
```

File: ci-operator/step-registry/network/conf/disable-cno/network-conf-disable-cno-ref.yaml
Comment (smarterclayton) [inline ci-operator/step-registry/network/conf/disable-cno/network-conf-disable-cno-ref.yaml:10]: +1

**Pull request merged**
